### ISSUE TITLE
Idea: Add a validation mechanism to confirm that an attack ran successfully

### ISSUE THREAD

**Issue by clr2of8** [redcanaryco/atomic-red-team#1194]
Often after running an atomic test it is unclear whether it ran successfully or not. It may have been blocked by AV or just didn't accomplish it's intended purpose. I would like to find a solution which would allow a user to easily validate that a test ran successfully. 

One proposed solution is to add the following optional yaml elements to the atomic test definition
```
validation_executor_name: <e.g. powershell>
validation_commands: <works like prereqs, must return 0 for success>
```
Users can optionally run the validation commands after a test run to check that a file, registry key, service etc exits or doesn't exist after the atomic test executes. The most straight forward validation would be ensuring a file, process, service, registry key, etc., were created.

**Comment by mgraeber-rc** [comment]
Thanks for posing this to the community, @clr2of8!

Pros of having separate validator logic:
* It allows for a clear distinction between code performing the procedure and code performing validation of the procedure.
* It could offer a motivation for people to separately implement validation logic.

Cons:
* Validation can already be performed in executor code potentially rendering separate validation logic redundant. e.g. the validation code in the [T1218.004 tests](https://github.com/redcanaryco/atomic-red-team/blob/24549e3866407c3080b95b6afebf78e8acd23352/atomics/T1218.004/T1218.004.yaml)
* It would likely miss an opportunity to validate ephemeral artifacts (e.g. short-lived processes).

I personally lean towards not having separate validation logic unless the downsides listed could be mitigated.

**Comment by mgraeber-rc** [comment]
Hey @clr2of8! Any additional thoughts on this? Are there any additional pros/cons I neglected to consider?

I agree that there should be a clearer way to specify the criteria for a passing test since that is largely ambiguous in most tests.

Having developed a handful of test harnesses in AtomicTestHarnesses now, depending on the technique, validation of pass criteria can sometimes be tricky involving several moving pieces and I've found it easiest to incorporate validation logic _with_ the test code.
